src/session_server.c

```diff
@@ -3080,3 +3098,754 @@ nc_session_get_notif_status(const struct nc_session *session)
+
+            /* update the next notification time */
+            curr_cert->notif_time = nc_server_notif_cert_exp_next_notif_time_get(intervals, interval_count, curr_cert);
+            if (curr_cert->notif_time == -1) {
+                break;
+            }
+        } else {
+            ERR(NULL, "Pthread condition timedwait failed (%s).", strerror(r));
+            break;
+        }
+    }
+
+cleanup:
+    VRB(NULL, "Certificate expiration notification thread exit.");
+    if (targ->clb_free_data) {
+        targ->clb_free_data(targ->clb_data);
+    }
+    nc_server_notif_cert_exp_dates_destroy(exp_dates, exp_date_count);
+    free(targ);
+    return NULL;
+}
+
+API int
+nc_server_notif_cert_expiration_thread_start(nc_cert_exp_notif_clb cert_exp_notif_clb,
+        void *user_data, void (*free_data)(void *))
+{
+    int r, ret = 0;
+    pthread_t tid;
+    struct nc_cert_exp_notif_thread_arg *arg;
+
+    NC_CHECK_ARG_RET(NULL, cert_exp_notif_clb, 1);
+
+    /* set the user callback and its data */
+    arg = malloc(sizeof *arg);
+    NC_CHECK_ERRMEM_RET(!arg, 1);
+    arg->clb = cert_exp_notif_clb;
+    arg->clb_data = user_data;
+    arg->clb_free_data = free_data;
+
+    /* LOCK */
+    pthread_mutex_lock(&server_opts.cert_exp_notif_thread_lock);
+
+    /* check if the thread is already running */
+    if (server_opts.cert_exp_notif_thread_running) {
+        ERR(NULL, "Certificate expiration notification thread is already running.");
+        ret = 1;
+        goto cleanup;
+    } else {
+        server_opts.cert_exp_notif_thread_running = 1;
+    }
+
+    if ((r = pthread_create(&tid, NULL, nc_server_notif_cert_exp_thread, arg))) {
+        ERR(NULL, "Creating the certificate expiration notification thread failed (%s).", strerror(r));
+        ret = 1;
+        goto cleanup;
+    }
+
+    server_opts.cert_exp_notif_thread_tid = tid;
+
+cleanup:
+    /* UNLOCK */
+    pthread_mutex_unlock(&server_opts.cert_exp_notif_thread_lock);
+    if (ret) {
+        free(arg);
+    }
+    return ret;
+}
+
+API void
+nc_server_notif_cert_expiration_thread_stop(int wait)
+{
+    int r;
+
+    /* LOCK */
+    pthread_mutex_lock(&server_opts.cert_exp_notif_thread_lock);
+    if (server_opts.cert_exp_notif_thread_running) {
+        /* set the running flag to 0, signal the thread and unlock its mutex */
+        server_opts.cert_exp_notif_thread_running = 0;
+        pthread_cond_signal(&server_opts.cert_exp_notif_thread_cond);
+
+        /* UNLOCK */
+        pthread_mutex_unlock(&server_opts.cert_exp_notif_thread_lock);
+        if (wait) {
+            r = pthread_join(server_opts.cert_exp_notif_thread_tid, NULL);
+            if (r) {
+                ERR(NULL, "Joining the certificate expiration notification thread failed (%s).", strerror(r));
+            }
+        }
+    } else {
+        /* thread is not running */
+        /* UNLOCK */
+        pthread_mutex_unlock(&server_opts.cert_exp_notif_thread_lock);
+    }
+}
```

Comment:
Not sure about this design. I wanted it to be blocking in the tests I ran, but someone may possibly not want to wait for the thread to terminate? Should the `wait` parameter be kept?